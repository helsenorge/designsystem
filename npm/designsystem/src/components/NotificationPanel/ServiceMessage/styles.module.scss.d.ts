export type Styles = {
  'service-message__bottom-row': string;
  'service-message__bottom-row__button': string;
  'service-message__bottom-row__close-button': string;
  'service-message__bottom-row__read-more-btn--spacing': string;
  'service-message__content': string;
  'service-message__content--smaller': string;
  'service-message__expander-button': string;
  'service-message__expander-button--close': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
