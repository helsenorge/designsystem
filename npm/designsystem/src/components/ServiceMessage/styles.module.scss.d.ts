export type Styles = {
  'service-message': string;
  'service-message__action': string;
  'service-message__action--close': string;
  'service-message__actions': string;
  'service-message__close': string;
  'service-message__col': string;
  'service-message__container': string;
  'service-message__content': string;
  'service-message__info': string;
  'service-message__info--extra': string;
  'service-message__label': string;
  'service-message__label-container': string;
  'service-message__label-container--has-expander': string;
  'service-message__row': string;
  'service-message__title': string;
  'service-message__toggle': string;
  'service-message--alert': string;
  'service-message--info': string;
  'service-message--success': string;
  'service-message--warn': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
