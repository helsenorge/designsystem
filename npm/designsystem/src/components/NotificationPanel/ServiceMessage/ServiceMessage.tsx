import React, { useRef } from 'react';
import classNames from 'classnames';

import styles from './styles.module.scss';

import Icon, { IconSize } from '../../Icons';
import NotificationPanel, { NotificationPanelProps } from '../NotificationPanel';
import { useHover } from '../../../hooks/useHover';
import { getColor } from '../../../theme/currys';
import X from '../../Icons/X';
import Forward from '../../Icons/Forward';

export type NotificationPanelVariants = 'info' | 'warn' | 'alert' | 'success';
export type NotificationCompactVariants = 'basic' | 'outline';
export type NotificationPanelSizes = 'small' | 'medium' | 'large' | 'full';

export interface ServiceMessageProps extends NotificationPanelProps {
  /** Sets a label for the notification panel. */
  serviceMessageLabel: string;
  /** function that runs on dismiss */
  onDismiss: () => void;
  /** Adds class to inner element. Alowing to pass in container width to the content only */
  innerElementClass?: string;
  /** Makes expander be open from start. */
  expanderOpenFromStart?: boolean;
  /** String displayed in service-message when expended*/
  serviceMessageInfo?: string;
  /** String displayed in service-message when expended, with a smaller font*/
  serviceMessageExtraInfo?: string;
  /** Name that describes url-link for the user*/
  serviceMessageReadMoreText?: string;
  /** Url for further information*/
  serviceMessageReadMoreUrl?: string;
  /** Text on close-expander button in service messages. */
  serviceMessageCloseBtnText?: string;
  /** Toggles if the message can be the closed. */
  dismissable?: boolean;
}

const ServiceMessage: React.FC<ServiceMessageProps> = (props: ServiceMessageProps) => {
  const {
    serviceMessageLabel,
    onDismiss,
    serviceMessageInfo,
    serviceMessageExtraInfo,
    serviceMessageReadMoreText,
    serviceMessageReadMoreUrl,
    serviceMessageCloseBtnText,
    expanderOpenFromStart = false,
    innerElementClass = 'container',
    dismissable = true,
  } = props;

  const anchorlinkClasses = classNames(
    styles['service-message__bottom-row__button'],
    styles['service-message__bottom-row__read-more-btn'],
    styles['service-message__bottom-row__read-more-btn--spacing']
  );
  const hasExpander = !!serviceMessageInfo || !!serviceMessageExtraInfo;
  const closeButtonClasses = classNames(styles['service-message__bottom-row__button'], styles['service-message__bottom-row__close-button']);

  const ButtonRow = () => {
    const readMoreRef = useRef<HTMLAnchorElement>(null);
    const { isHovered: readMoreHoverRefIsHovered } = useHover(readMoreRef);
    const xRef = useRef<HTMLButtonElement>(null);
    const { isHovered: xRefIsHovered } = useHover(xRef);
    return (
      <div className={styles['service-message__bottom-row']}>
        <a className={anchorlinkClasses} href={serviceMessageReadMoreUrl} ref={readMoreRef}>
          {serviceMessageReadMoreText}
          <Icon size={IconSize.XSmall} svgIcon={Forward} color={getColor('blueberry', 700)} isHovered={readMoreHoverRefIsHovered} />
        </a>

        <button ref={xRef} className={closeButtonClasses} aria-label={serviceMessageCloseBtnText} onClick={onDismiss}>
          {serviceMessageCloseBtnText}
          <Icon size={IconSize.XSmall} svgIcon={X} color={getColor('blueberry', 700)} isHovered={xRefIsHovered} />
        </button>
      </div>
    );
  };

  return (
    <NotificationPanel
      {...props}
      label={serviceMessageLabel}
      expanderOpenFromStart={expanderOpenFromStart}
      dismissable={dismissable}
      isServiceMessage
      contentClassName={innerElementClass}
      serviceMessageExpander={hasExpander}
    >
      {hasExpander && (
        <div>
          {serviceMessageInfo && <div className={styles['service-message__content']}>{serviceMessageInfo}</div>}
          {serviceMessageExtraInfo && <div className={styles['service-message__content--smaller']}>{serviceMessageExtraInfo}</div>}

          <ButtonRow />
        </div>
      )}
    </NotificationPanel>
  );
};

export default ServiceMessage;
