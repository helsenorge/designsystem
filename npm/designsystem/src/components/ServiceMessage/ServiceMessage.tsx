import React, { useState, useRef } from 'react';

import classNames from 'classnames';

import { useBreakpoint } from '../../hooks/useBreakpoint';
import { useHover } from '../../hooks/useHover';
import { useUuid } from '../../hooks/useUuid';
import { getColor } from '../../theme/currys';
import { breakpoints } from '../../theme/grid';
import { palette } from '../../theme/palette';
import { getAriaLabelAttributes } from '../../utils/accessibility';
import { AnchorLinkTargets } from '../AnchorLink';
import Close from '../Close';
import Icon, { IconSize } from '../Icons';
import CheckFill from '../Icons/CheckFill';
import ChevronDown from '../Icons/ChevronDown';
import ChevronUp from '../Icons/ChevronUp';
import ErrorSignFill from '../Icons/ErrorSignFill';
import Forward from '../Icons/Forward';
import InfoSignFill from '../Icons/InfoSignFill';
import TriangleX from '../Icons/TriangleX';
import X from '../Icons/X';
import { NotificationPanelVariants } from '../NotificationPanel';

import styles from './styles.module.scss';

interface LabelProps {
  label: string;
  variant: NotificationPanelVariants;
  id: string;
  hasExpander: boolean;
  isExpanded: boolean;
  dismissable: boolean;
  onExpand: () => void;
  onDismiss?: () => void;
  closeBtnText?: string;
}

const Label: React.FC<LabelProps> = ({ label, variant, id, hasExpander, isExpanded, dismissable, onExpand, onDismiss, closeBtnText }) => {
  const breakpoint = useBreakpoint();
  const { isHovered, hoverRef } = useHover<HTMLDivElement>();

  const iconSize = breakpoint < breakpoints.lg ? IconSize.XSmall : IconSize.Small;
  const variantToIconMap = {
    info: (
      <Icon svgIcon={InfoSignFill} color={palette.blueberry700} hoverColor={palette.blueberry700} size={iconSize} isHovered={isHovered} />
    ),
    warn: <Icon svgIcon={ErrorSignFill} color={palette.banana700} hoverColor={palette.banana700} size={iconSize} isHovered={isHovered} />,
    alert: <Icon svgIcon={TriangleX} color={palette.cherry700} hoverColor={palette.cherry700} size={iconSize} isHovered={isHovered} />,
    success: <Icon svgIcon={CheckFill} color={palette.kiwi900} hoverColor={palette.kiwi900} size={iconSize} isHovered={isHovered} />,
  };
  const CustomTag = hasExpander ? 'button' : 'span';

  const labelContainerClasses = classNames(
    styles['service-message__label-container'],
    styles[`service-message__label-container--${variant}`],
    hasExpander && styles[`service-message__label-container--has-expander`]
  );

  return (
    <div className={labelContainerClasses} ref={hoverRef}>
      <div className={styles['service-message__container']}>
        <div className={styles['service-message__row']}>
          <div className={styles['service-message__col']}>
            <div className={styles['service-message__label']}>
              {variantToIconMap[variant]}
              <h1 className={styles['service-message__title']} id={id}>
                <CustomTag
                  className={styles['service-message__toggle']}
                  onClick={hasExpander ? onExpand : undefined}
                  aria-expanded={hasExpander ? isExpanded : undefined}
                >
                  {label}
                </CustomTag>
              </h1>
              {hasExpander && <Icon size={iconSize} svgIcon={isExpanded ? ChevronUp : ChevronDown} isHovered={isHovered} />}
              {!hasExpander && dismissable && (
                <Close onClick={onDismiss} ariaLabel={closeBtnText} className={styles['service-message__close']} />
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

interface ContentProps {
  info?: string;
  extraInfo?: string;
  urlTitle?: string;
  url?: string;
  target?: AnchorLinkTargets;
  dismissable: boolean;
  onDismiss?: () => void;
  closeBtnText?: string;
}

const Content: React.FC<ContentProps> = ({ info, extraInfo, urlTitle, url, target, dismissable, closeBtnText, onDismiss }) => {
  const { hoverRef: readMoreRef, isHovered: readMoreHoverRefIsHovered } = useHover<HTMLAnchorElement>();
  const { hoverRef: closeButtonRef, isHovered: closeButtonIsHovered } = useHover<HTMLButtonElement>();

  const hasUrl = url && urlTitle;

  return (
    <div className={styles['service-message__container']}>
      <div className={styles['service-message__row']}>
        <div className={styles['service-message__col']}>
          <div className={styles['service-message__content']}>
            {info && <p className={styles['service-message__info']}>{info}</p>}
            {extraInfo && (
              <p className={classNames(styles['service-message__info'], styles['service-message__info--extra'])}>{extraInfo}</p>
            )}
            <div className={styles['service-message__actions']}>
              {hasUrl && (
                <a className={styles['service-message__action']} href={url} ref={readMoreRef} target={target}>
                  {urlTitle}
                  <Icon size={IconSize.XSmall} svgIcon={Forward} color={getColor('blueberry', 700)} isHovered={readMoreHoverRefIsHovered} />
                </a>
              )}

              {dismissable && (
                <button
                  ref={closeButtonRef}
                  className={classNames(styles['service-message__action'], styles['service-message__action--close'])}
                  aria-label={closeBtnText}
                  onClick={onDismiss}
                >
                  {closeBtnText}
                  <Icon size={IconSize.XSmall} svgIcon={X} color={getColor('blueberry', 700)} isHovered={closeButtonIsHovered} />
                </button>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export interface ServiceMessageProps {
  /** Sets a label for the notification panel. */
  label: string;
  /** String displayed in service-message when expanded */
  info?: string;
  /** String displayed in service-message when expanded, with a smaller font */
  extraInfo?: string;
  /** function that runs on dismiss */
  onDismiss?: () => void;
  /** Allows for dismiss to be an option, also for ServiceMessage with only label as content */
  dismissable?: boolean;
  /** Makes expander be open from start. */
  expanderOpenFromStart?: boolean;
  /** Name that describes a url-link for the user*/
  urlTitle?: string;
  /** Url for further information*/
  url?: string;
  /** Sets target for the anchorlink to the url*/
  target?: AnchorLinkTargets;
  /** Text on close-button in service messages. */
  closeBtnText?: string;
  /** Changes the visual representation. */
  variant?: NotificationPanelVariants;
  /** Sets the data-testid attribute. */
  testId?: string;
  /**
   * First of its type. Used to remove border-top
   * @deprecated Har ingen effekt p√• komponenten
   */
  first?: boolean;
}

const ServiceMessage: React.FC<ServiceMessageProps> = ({
  label,
  dismissable = true,
  onDismiss,
  info,
  extraInfo,
  urlTitle,
  url,
  target = '_self',
  closeBtnText = 'fjern melding',
  expanderOpenFromStart = false,
  variant = 'alert',
  testId,
}) => {
  const [isExpanded, setIsExpanded] = useState<boolean>(expanderOpenFromStart);

  const labelId = useUuid();
  const hasExpander = !!info || !!extraInfo;

  const ariaRole = variant === 'alert' ? 'alert' : 'region';
  const ariaLabelAttributes = getAriaLabelAttributes({ label, id: labelId });

  const handleClick = (): void => {
    hasExpander && setIsExpanded(!isExpanded);
  };

  const classes = classNames(
    styles['service-message'],
    styles[`service-message--${variant}`],
    isExpanded && styles[`service-message--expanded`]
  );

  return (
    <div className={classes} role={ariaRole} {...ariaLabelAttributes} data-testid={testId}>
      <Label
        label={label}
        variant={variant}
        id={labelId}
        hasExpander={hasExpander}
        isExpanded={isExpanded}
        dismissable={dismissable}
        onExpand={handleClick}
        onDismiss={onDismiss}
        closeBtnText={closeBtnText}
      />
      {hasExpander && isExpanded && (
        <Content
          info={info}
          extraInfo={extraInfo}
          urlTitle={urlTitle}
          url={url}
          target={target}
          dismissable={dismissable}
          onDismiss={onDismiss}
          closeBtnText={closeBtnText}
        />
      )}
    </div>
  );
};

export default ServiceMessage;
