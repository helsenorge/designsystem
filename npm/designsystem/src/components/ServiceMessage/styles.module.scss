@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/palette';
@import '../../scss/breakpoints';
@import '../../scss/font-settings';
@import '../../scss/grid';

.service-message {
  $servicemessage: &;

  border-width: 1px;
  border-bottom-style: solid;

  &:first-of-type {
    border-top-style: solid;
  }

  &--success {
    background-color: $kiwi50;

    &:hover {
      background-color: rgba-to-rgb(rgba($kiwi500, 0.1), $kiwi50);
    }

    &#{$servicemessage},
    &#{$servicemessage}__label-container {
      border-color: $kiwi900;
    }
  }

  &--info {
    background-color: $blueberry50;

    &:hover {
      background-color: rgba-to-rgb(rgba($blueberry500, 0.1), $blueberry50);
    }

    &#{$servicemessage},
    &#{$servicemessage}__label-container {
      border-color: $blueberry700;
    }
  }

  &--warn {
    background-color: $banana50;

    &:hover {
      background-color: rgba-to-rgb(rgba($banana500, 0.1), $banana50);
    }

    &#{$servicemessage},
    &#{$servicemessage}__label-container {
      border-color: $banana700;
    }
  }

  &--alert {
    background-color: $cherry50;

    &:hover {
      background-color: rgba-to-rgb(rgba($cherry500, 0.1), $cherry50);
    }

    &#{$servicemessage},
    &#{$servicemessage}__label-container {
      border-color: $cherry700;
    }
  }

  &__container {
    @include make-container;
    @include make-container-max-widths;
  }

  &__row {
    @include make-row;
  }

  &__col {
    @include make-col-ready;
    @include make-col(12);
  }

  &__label-container {
    position: relative;
    padding-top: getSpacer(3xs);
    padding-bottom: getSpacer(3xs);

    @media (min-width: map-get($grid-breakpoints, lg)) {
      padding-top: getSpacer(2xs);
      padding-bottom: getSpacer(2xs);
    }

    &--has-expander:has(:focus-visible) {
      box-shadow: inset 0 0 0 2px $black;
    }
  }

  &__label {
    display: flex;
    width: 100%;
    align-items: center;
    gap: getSpacer(2xs);

    @media (min-width: map-get($grid-breakpoints, md)) {
      gap: getSpacer(m);
    }
  }

  &__title {
    width: 100%;
    font-weight: 700;
    font-size: $font-size-xs;
    line-height: 1.25rem;

    @media (min-width: map-get($grid-breakpoints, md)) {
      font-size: $font-size-sm;
    }
  }

  &__toggle {
    all: unset;

    // Triks for at hele labelen skal trigge klikk p√• knappen
    &::after {
      inset: 0;
      content: '';
      position: absolute;
    }
  }

  &__content {
    margin-left: calc(38px + getSpacer(2xs));
    margin-right: calc(38px + getSpacer(2xs));
    padding-top: getSpacer(xs);
    padding-bottom: getSpacer(xs);

    @media (min-width: map-get($grid-breakpoints, md)) {
      margin-left: calc(38px + getSpacer(m));
      margin-right: calc(38px + getSpacer(m));
    }

    @media (min-width: map-get($grid-breakpoints, lg)) {
      margin-left: calc(48px + getSpacer(m));
      margin-right: calc(48px + getSpacer(m));
      padding-top: getSpacer(2xs);
      padding-bottom: getSpacer(s);
    }
  }

  &__info {
    font-size: $font-size-xs;
    line-height: 1.25rem;
    margin: 0;

    @media (min-width: map-get($grid-breakpoints, lg)) {
      margin-top: getSpacer(xs);
      line-height: 1.625rem;
      font-size: $font-size-sm;
    }

    &--extra {
      font-size: 0.875rem;
      margin-top: getSpacer(2xs);

      @media (min-width: map-get($grid-breakpoints, lg)) {
        margin-top: getSpacer(xs);
        font-size: $font-size-xs;
        line-height: $lineheight-size-xs;
      }
    }
  }

  &__actions {
    display: flex;
    justify-content: space-between;
    flex-flow: column wrap;
    gap: getSpacer(2xs);
    margin-top: getSpacer(2xs);

    @media (min-width: map-get($grid-breakpoints, sm)) {
      flex-direction: row;
    }

    @media (min-width: map-get($grid-breakpoints, lg)) {
      margin-top: getSpacer(s);
    }
  }

  &__action {
    display: flex;
    align-items: center;
    border: none;
    width: fit-content;
    color: $blueberry700;
    font-weight: 600;
    font-size: $font-size-xs;
    line-height: 1.25rem;
    background-color: transparent;
    text-decoration: none;
    padding: 0.5rem;
    margin-left: -0.5rem;
    gap: getSpacer(xs);

    &:hover,
    &:active {
      background-color: rgb(88 170 187 / 10%);
    }

    &:focus-visible {
      box-shadow: 0 0 0 2px $black;
      background-color: transparent;
      border: 0;
      outline: none;
    }

    &--close {
      @media (min-width: map-get($grid-breakpoints, lg)) {
        margin-left: auto;
      }
    }
  }

  &__close {
    z-index: 1;
  }
}
