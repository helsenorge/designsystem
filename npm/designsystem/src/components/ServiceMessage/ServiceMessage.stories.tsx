import React from 'react';

import { action } from '@storybook/addon-actions';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import ServiceMessage from './ServiceMessage';

export default {
  title: 'Components/ServiceMessage',
  component: ServiceMessage,
  parameters: {
    docs: {
      description: {
        component:
          'ServiceMessage lar brukeren velge et av flere valg i en liste. ServiceMessage kan brukes frittstående, som en del av en FormGroup eller direkte i et Validation komponent.',
      },
    },
  },

  argTypes: {
    label: {
      control: 'text',
      defaultValue: 'ServiceMessage label',
    },
    info: {
      control: 'text',
      defaultValue: 'Nedetid for østre-aker',
    },
    extraInfo: {
      control: 'text',
      defaultValue: 'Ring 08400 for mer informasjon',
    },
    dismissable: {
      control: 'boolean',
      defaultValue: true,
    },
    expanderOpenFromStart: {
      control: 'boolean',
      defaultValue: true,
    },
    urlTitle: {
      control: 'text',
      defaultValue: 'Du kan henvises fra sykehuset eller fra din fastlege (helsedirektoratet.no)',
    },
    url: {
      control: 'text',
      defaultValue: '/',
    },
    target: {
      control: 'select',
      options: ['_self', '_blank', '_parent'],
      defaultValue: '_self',
    },
    closeBtnText: {
      control: 'text',
      defaultValue: 'fjern melding',
    },
    variant: {
      control: 'select',
      options: ['info', 'warn', 'alert', 'success'],
      defaultValue: 'alert',
    },
    testId: {
      control: 'text',
      defaultValue: 'test',
    },
    first: {
      control: 'boolean',
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof ServiceMessage>;

export const Default: ComponentStory<typeof ServiceMessage> = (args: any) => (
  <>
    <ServiceMessage {...args} />
    <ServiceMessage {...args} expanderOpenFromStart={false} />
    <ServiceMessage {...args} expanderOpenFromStart={false} variant="info" />
    <ServiceMessage {...args} expanderOpenFromStart={false} variant="success" />
    <ServiceMessage {...args} expanderOpenFromStart={false} variant="warn" />
  </>
);
export const ReadMore: ComponentStory<typeof ServiceMessage> = (args: any) => <ServiceMessage {...args} dismissable={false} />;
export const DismissableWithContent: ComponentStory<typeof ServiceMessage> = (args: any) => (
  <ServiceMessage
    label={'Dismissable with content'}
    info="You can dismiss!"
    extraInfo="If you like"
    url={undefined}
    urlTitle={undefined}
    dismissable={true}
    onDismiss={action('Dismiss clicked')}
  />
);
export const DismissableLabelOnly: ComponentStory<typeof ServiceMessage> = (args: any) => (
  <ServiceMessage label={'Dismissable label only'} variant="alert" dismissable={true} onDismiss={action('Dismiss clicked')} />
);

export const LabelOnly: ComponentStory<typeof ServiceMessage> = (args: any) => (
  <ServiceMessage label={'Label only'} variant="alert" dismissable={false} />
);
