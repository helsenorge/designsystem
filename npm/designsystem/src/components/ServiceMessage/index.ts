import ServiceMessage from './ServiceMessage';
export * from './ServiceMessage';
export default ServiceMessage;
