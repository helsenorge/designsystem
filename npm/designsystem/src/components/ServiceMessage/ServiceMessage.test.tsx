import React from 'react';

import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ServiceMessage from './ServiceMessage';

describe('Gitt at ServiceMessage skal vises', () => {
  describe('Når komponenten vises med default props', () => {
    test('Så rendres komponenten riktig', (): void => {
      const { container } = render(<ServiceMessage label={'Some text here for testing.'} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når komponenten skal kunne lukkes', () => {
    test('Så kalles onClick-handler når man klikker på lukkeknappen', async (): Promise<void> => {
      const mockClose = jest.fn();

      render(
        <ServiceMessage label={'Some text here for testing.'} onDismiss={mockClose} dismissable closeBtnText="Lukk denne beskjeden" />
      );

      const closeButton = screen.getByRole('button', { name: 'Lukk denne beskjeden' });

      await userEvent.click(closeButton);
      expect(mockClose).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når testId-prop er satt', (): void => {
    test('Så kan komponenten finnes ved hjelp av testId', (): void => {
      render(<ServiceMessage testId="bare-tester" label={'Some text here for testing.'} />);

      const component = screen.getByTestId('bare-tester');
      expect(component).toBeVisible();
    });
  });

  describe('Når komponenten ikke har expander og bruker klikker på tab', (): void => {
    test('Så får lukkeknappen fokus først', async (): Promise<void> => {
      const mockClose = jest.fn();
      render(<ServiceMessage label={'Some text here for testing.'} dismissable closeBtnText="lukk" onDismiss={mockClose} />);

      await userEvent.tab();

      const closeButton = screen.getByRole('button', { name: 'lukk' });

      expect(closeButton).toHaveFocus();
    });
  });
  describe('Når komponenten har expander og bruker klikker på tab', (): void => {
    test('Så får øverste rad/knapp fokus først', async (): Promise<void> => {
      const mockClose = jest.fn();
      render(<ServiceMessage label={'Some text here for testing.'} dismissable closeBtnText="lukk" info="hei" onDismiss={mockClose} />);

      await userEvent.tab();

      const button = screen.getByRole('button', { name: 'Some text here for testing.' });

      expect(button).toHaveFocus();
    });
  });

  describe('Når bruker klikker på expand', (): void => {
    test('Så vises innholdet i expander', async (): Promise<void> => {
      render(<ServiceMessage label={'Some text here for testing.'} info={'hei'} />);

      const expandButton = screen.getByRole('button', { name: 'Some text here for testing.' });
      await userEvent.click(expandButton);

      const info = screen.getByText('hei');
      expect(info).toBeVisible();
    });
  });
  describe('Når expander skal være åpen fra start', (): void => {
    test('Så vises innholdet i expander fra start', async (): Promise<void> => {
      render(<ServiceMessage expanderOpenFromStart label={'Some text here for testing.'} info={'hei'} />);

      const info = screen.getByText('hei');
      expect(info).toBeVisible();
    });
  });
  describe('Når komponenten har expander', () => {
    test('Så kalles onClick-handler når man klikker på lukkeknappen etter å ha ekspandert', async (): Promise<void> => {
      const mockClose = jest.fn();
      render(
        <ServiceMessage
          label={'Some text here for testing.'}
          info={'hei'}
          closeBtnText="Lukk denne beskjeden"
          onDismiss={mockClose}
          dismissable
        />
      );

      const expandButton = screen.getByRole('button', { name: 'Some text here for testing.' });
      await userEvent.click(expandButton);
      const closeButton = screen.getByRole('button', { name: 'Lukk denne beskjeden' });
      await userEvent.click(closeButton);
      expect(mockClose).toHaveBeenCalledTimes(1);
    });
  });
});
