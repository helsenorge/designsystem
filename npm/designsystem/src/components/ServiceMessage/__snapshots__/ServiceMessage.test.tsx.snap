// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at ServiceMessage skal vises Når komponenten vises med default props Så rendres komponenten riktig 1`] = `
<div>
  <div
    aria-labelledby="test"
    class="service-message service-message--alert"
    role="alert"
  >
    <div
      class="service-message__label-container service-message__label-container--alert"
    >
      <div
        class="service-message__container"
      >
        <div
          class="service-message__row"
        >
          <div
            class="service-message__col"
          >
            <div
              class="service-message__label"
            >
              <svg
                aria-hidden="true"
                class="hnds-style-icon"
                data-analyticsid="icon"
                fill="#A31F0E"
                focusable="false"
                height="48"
                role="presentation"
                style="min-width: 48px; min-height: 48px;"
                viewBox="0 0 48 48"
                width="48"
              >
                <path
                  clip-rule="evenodd"
                  d="m24 7 19 33.712H5L24 7Zm-5.792 17.951a1.185 1.185 0 1 1 1.676-1.675l4.01 4.01 3.895-3.895a1.19 1.19 0 0 1 1.683 1.683l-3.895 3.895 4.387 4.387a1.185 1.185 0 0 1-1.675 1.675l-4.387-4.386-4.503 4.502a1.19 1.19 0 0 1-1.683-1.683l4.503-4.502-4.01-4.011Z"
                  fill-rule="evenodd"
                />
              </svg>
              <h1
                class="service-message__title"
                id="test"
              >
                <span
                  class="service-message__toggle"
                >
                  Some text here for testing.
                </span>
              </h1>
              <button
                aria-label="fjern melding"
                class="close service-message__close"
                data-analyticsid="close"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="hnds-style-icon"
                  data-analyticsid="icon"
                  fill="#126F87"
                  focusable="false"
                  height="48"
                  role="presentation"
                  style="min-width: 48px; min-height: 48px;"
                  viewBox="0 0 48 48"
                  width="48"
                >
                  <path
                    d="M25.403 24l10.259-10.259-1.403-1.403L24 22.597l-10.259-10.26-1.403 1.403L22.597 24 12.338 34.26l1.403 1.403L24 25.403l10.259 10.259 1.403-1.403z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
